### PR TITLE
add text spotting demo

### PR DESCRIPTION
**Pull request by Ilya-Krylov**

### CODE REVIEW

**Comment by Ilya-Krylov** [comment]
@IRDonch

**Comment by Ilya-Krylov** [comment]
@Wovchena could you please have a look?

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+            input_image = cv2.resize(frame, None, fx=scale_x, fy=scale_y)
+
+        input_image_size = input_image.shape[:2]
+        input_image = np.pad(input_image, ((0, h - input_image_size[0]),
+                                           (0, w - input_image_size[1]),
+                                           (0, 0)),
+                             mode='constant', constant_values=0)
+        # Change data layout from HWC to CHW.
+        input_image = input_image.transpose((2, 0, 1))
+        input_image = input_image.reshape((n, c, h, w)).astype(np.float32)
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: actually it expands dims adding batch dimension

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+            scale_y = h / frame.shape[0]
+            input_image = cv2.resize(frame, (w, h))
+        else:
+            # Resize the image to keep the same aspect ratio and to fit it to a window of a target size.
+            scale_x = scale_y = min(h / frame.shape[0], w / frame.shape[1])
+            input_image = cv2.resize(frame, None, fx=scale_x, fy=scale_y)
+
+        input_image_size = input_image.shape[:2]
+        input_image = np.pad(input_image, ((0, h - input_image_size[0]),
+                                           (0, w - input_image_size[1]),
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: it does not matter, this is made as in original instance_segmentation_demo.py.

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+        text_features = outputs['text_features']
+
+        # Filter out detections with low confidence.
+        detections_filter = scores > args.prob_threshold
+        scores = scores[detections_filter]
+        classes = classes[detections_filter]
+        boxes = boxes[detections_filter]
+        masks = list(segm for segm, is_valid in zip(masks, detections_filter) if is_valid)
+        text_features = text_features[detections_filter]
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: what do u mean? If you are talking about 'text_features', then they are filtered first and only than decoded

[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: got it, thanks

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: OK

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: Thank, you. I didn't notice that.

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: `boxes` are scaled first and than filtered. `masks` are postprocessed first and that filtered by `args.prob_threshold`. My suggestion is to filter them first and postprocess what's left.

I should have excluded the last line (`text_features = text_features[detections_filter]`) from comment. I think there is nothing to update about `text_features`.

File: demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                              [218, 218, 218],
+                              [255, 255, 255]], dtype=np.uint8)
+
+    def __init__(self, class_labels, confidence_threshold=0.5, show_boxes=False,
+                 show_masks=True, show_scores=False):
+        super().__init__()
+        self.class_labels = class_labels
+        self.confidence_threshold = confidence_threshold
+        self.class_color_palette = np.asarray([2 ** 25 - 1, 2 ** 15 - 1, 2 ** 21 - 1])
+        self.instance_color_palette = self.color_palette
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py]: done

File: demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        for box, color in zip(boxes, colors):
+            box = box.astype(int)
+            top_left, bottom_right = box[:2].tolist(), box[2:].tolist()
+            image = cv2.rectangle(
+                image, tuple(top_left), tuple(bottom_right), tuple(color), 1
+            )
+        return image
+
+    def overlay_masks(self, image, masks, classes, ids=None):
+        colors = self.compute_colors_for_labels(classes).tolist()
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py]: done

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+        # Get instance track IDs.
+        masks_tracks_ids = None
+        if tracker is not None:
+            masks_tracks_ids = tracker(masks, classes)
+
+        # Visualize masks.
+        frame = visualizer(frame, boxes, classes, scores, masks, texts, masks_tracks_ids)
+
+        # Draw performance stats.
+        inf_time_message = 'Inference time: {:.3f} ms'.format(det_time * 1000)
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: done

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+            feature = text_enc_exec_net.infer({'input': feature})['output']
+            feature = np.reshape(feature, (feature.shape[0], feature.shape[1], -1))
+            feature = np.transpose(feature, (0, 2, 1))
+
+
+            hidden = np.zeros(hidden_shape)
+            prev_symbol_index = np.ones((1,)) * SOS_INDEX
+
+            vocab_size = len(args.alphabet)
+            per_feature_outputs = np.zeros([MAX_SEQ_LEN, vocab_size])
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: done

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: done

File: demos/python_demos/text_spotting_demo/README.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+  --no_show             Optional. Don't show output
+```
+
+Running the application with an empty list of options yields the short version of the usage message and an error message.
+
+To run the demo, you can use public or pre-trained models. To download the pre-trained models, use the OpenVINO [Model Downloader](../../../tools/downloader/README.md) or go to [https://download.01.org/opencv/](https://download.01.org/opencv/).
+
+> **NOTE**: Before running the demo with a trained model, make sure the model is converted to the Inference Engine format (`*.xml` + `*.bin`) using the [Model Optimizer tool](https://docs.openvinotoolkit.org/latest/_docs_MO_DG_Deep_Learning_Model_Optimizer_DevGuide.html).
+
+To run the demo, please provide paths to the model in the IR format, to a file with class labels, and to an input video, image, or folder with images:
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/README.md]: done

File: demos/python_demos/text_spotting_demo/README.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    * `classes` with object class IDs for all bounding boxes
+    * `raw_masks` with fixed-size segmentation heat maps for all classes of all bounding boxes
+    * `text_features` with text features which are feed to Text Recognition Head further
+
+Second model is Text Recognition Encoder that takes `text_features` as input and produces `encoded text`.
+
+Third model is Text Recognition Decoder that takes `encoded text` from Text Recognition Encoder ,`previous symbol` and `hidden state`. On the first step special `Start Of Sequence (SOS)` symbol and zero `hidden state` are fed to Text Recognition Decoder. The decoder produces `symbols distribution`, `current hidden state` each step until `End Of Sequence (EOS)` symbol is generated.
+
+As input, the demo application takes:
+* a path to a single image file, a video file or a numeric ID of a web camera specified with a command-line argument `-i`
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/README.md]: done

File: demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+                              [0, 0, 0],
+                              [36, 36, 36],
+                              [72, 72, 72],
+                              [109, 109, 109],
+                              [145, 145, 145],
+                              [182, 182, 182],
+                              [218, 218, 218],
+                              [255, 255, 255]], dtype=np.uint8)
+
+    def __init__(self, class_labels, confidence_threshold=0.5, show_boxes=False,
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py]: done

File: demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+        self.class_labels = class_labels
+        self.confidence_threshold = confidence_threshold
+        self.class_color_palette = np.asarray([2 ** 25 - 1, 2 ** 15 - 1, 2 ** 21 - 1])
+        self.instance_color_palette = self.color_palette
+        self.show_masks = show_masks
+        self.show_boxes = show_boxes
+        self.show_scores = show_scores
+
+    def __call__(self, image, boxes, classes, scores, segms, texts, ids=None):
+        result = image.copy()
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py]: done

**Comment by Ilya-Krylov** [comment]
@Wovchena thank you for your review, could you have a look one more time

File: demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Wovchena) [reply demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py]: `result = image` is confusing. I think you should just rename all appearances of `result` in the function.

[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo/visualizer.py]: done

**Comment by Wovchena** [comment]
Please, also update the list of demos in demos root README

**Comment by Ilya-Krylov** [comment]
@Wovchena done

**Comment by Ilya-Krylov** [comment]
@Wovchena there will be :)

File: demos/python_demos/text_spotting_demo/README.md [hunk 12]
```diff
@@ -0,0 +1,132 @@
+# Text Spotting Python* Demo
+
+This demo shows how to run Text Spotting models using OpenVINO&trade;.
```
Comment (IRDonch) [inline demos/python_demos/text_spotting_demo/README.md]: This is somewhat tautological. It would be nice if you could explain what text spotting actually _is_.

File: demos/python_demos/text_spotting_demo/text_spotting_demo.py [hunk 13]
```diff
@@ -0,0 +1,386 @@
...
+        if not args.no_show:
+            key = cv2.waitKey(args.delay)
+            esc_code = 27
+            if key == esc_code:
+                break
+
+    cv2.destroyAllWindows()
+    cap.release()
+    del mask_rcnn_exec_net
+    del ie
```
[the remark this answers is not in the document]
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/text_spotting_demo.py]: removed

File: demos/python_demos/text_spotting_demo/README.md (hunk 12 above)
Comment (Ilya-Krylov) [reply demos/python_demos/text_spotting_demo/README.md]: added one more sentence

File: demos/python_demos/text_spotting_demo/README.md [hunk 14]
```diff
@@ -0,0 +1,132 @@
+# Text Spotting Python* Demo
+
+This demo shows how to run Text Spotting models. Text Spotting models allows us to simultaneously detect and recognize text.
```
Comment (IRDonch) [inline demos/python_demos/text_spotting_demo/README.md]:
```suggestion
This demo shows how to run Text Spotting models. Text Spotting models allow us to simultaneously detect and recognize text.
```

**Comment by IRDonch** [comment]
Jenkins: test this please
